### DISCUSSION TITLE
Some packages have an installation script

### DISCUSSION THREAD

**Discussion by vercotux** [outlandishideas/wpackagist#130] [Ideas]
Some packages in wpackagist do not work because they require initialization code to be run on installation (usually to setup database tables). An example of such plugin is [anspress](https://wpackagist.org/search?q=anspress-question-answer). This is a MAJOR problem. Have you guys considered installing these plugins/themes using wp-cli?

**Comment by tamlyn** [comment]
Composer just puts the files in the right place (just as you would if you were uploading the files via FTP). You then need to activate the plugin either via the admin interface or using WP-CLI. It is on activation that DB tables are usually set up.

**Comment by vercotux** [comment]
> Composer just puts the files in the right place

That's not true. Composer supports [custom installers](https://getcomposer.org/doc/articles/custom-installers.md) which can be used to run a wp-cli command. I have done this myself.

EDIT: To clarify, it does not seem likely that DB tables are always set up on actiavtion. I have stumbled across at least 3 different plugins which did not work by simply putting the files in the right place (upon activation it gave me a [white screen of death](http://codex.wordpress.org/Common_WordPress_Errors#The_White_Screen_of_Death)). This seems to be even more common with non-free (paid) plugins.

**Comment by tamlyn** [comment]
If you manually activate the plugin after Composer installs it, does it not work? I understand that Composer could be configured to activate plugins after install but I'm not sure this should be the default behaviour. 

If you would like to contribute some documentation on how to achieve this using Composer hooks that would be very much appreciated.

**Comment by vercotux** [comment]
> If you manually activate the plugin after Composer installs it, does it not work?

Nope. Activation itself is not enough. The critical part happens on installation. That is why wp-cli provides both [activate](http://wp-cli.org/commands/plugin/activate/) and [install](http://wp-cli.org/commands/plugin/install/). Indeed, Composer should not be activating plugins/themes, but it should include them in a way that they are actually usable, and one way to do that is to rely on wp-cli's install function.

**Comment by tamlyn** [comment]
I'm going to have to do some testing to understand what you mean but it won't be before next week. Feel free to nudge me if I forget.

**Comment by vercotux** [comment]
_nudge_

**Comment by emilbayes** [comment]
I've looked at the Anspress plugin and don't see where the installation part lies. I'm trying to understand the difference between installation and activation. What I gather from skimming the `wp-cli` source is that installation is ~~`wp plugin get && wp plugin activate`~~ `curl PLUGIN_URL | tar zxf -C PLUGIN_DIR - && wp plugin activate` while `wp plugin activate` assumes that the plugin is already in the `plugins` directory. Is this right @vercotux?

**EDIT**: For completeness, here's the files I looked at in `wp-cli`:

[`wp plugin install`](https://github.com/wp-cli/wp-cli/blob/ebf01536febd52ad7a15654a19dc392c9ab6b97b/php/commands/plugin.php#L439-L446) executes the `parent::install` function, which downloads the plugin and activates it. [`parent::install`](https://github.com/wp-cli/wp-cli/blob/ebf01536febd52ad7a15654a19dc392c9ab6b97b/php/WP_CLI/CommandWithUpgrade.php#L152-L162) (highlighting what's important).

[`wp plugin activate`](https://github.com/wp-cli/wp-cli/blob/ebf01536febd52ad7a15654a19dc392c9ab6b97b/php/commands/plugin.php#L182) does a bunch of bookkeeping but eventually calls the Wordpress [`activate_plugin`](https://codex.wordpress.org/Function_Reference/activate_plugin) function.

**EDIT 2**: `wp plugin get` simply lists the information about a plugin

**Comment by NoelLH** [comment]
I've converted this to a Discussion for now as I'm not completely clear on what the enhancement would look like. Please feel free to create a detail _enhancement_ issue with a story on exactly how this should work once we know.
